### PR TITLE
Use WithProperties to improve properties handling (CASSANDRA-18453)

### PR DESCRIPTION
**Pull request by bbotella**
Using WithProperties to improve handling of property set and cleanups.

To find the actual try catches, I've used IntelliJ structural search with this pattern:

```
try ($ResourceType$ $resource$ = $init$; $expression$) {
  $TryStatement$;
} finally {
    $property$.clearValue();
}
```

patch by Bernardo Botella

### CODE REVIEW

**Comment by yifan-c** [comment]
Awesome to see the usage of structural search. We talked offline that the template needs to be improves to cover different code structures.

**Comment by bbotella** [comment]
> Awesome to see the usage of structural search. We talked offline that the template needs to be improves to cover different code structures.

I ended up going with Stephan's suggestion and spotted a bunch more cases.

File: checkstyle_test.xml [hunk 1]
```diff
@@ -118,6 +118,13 @@
       <property name="message" value="Use the CassandraRelevantProperties or CassandraRelevantEnv instead." />
     </module>
 
+    <module name="RegexpSinglelineJava">
+      <property name="id" value="clearValueSystemPropertyUsage"/>
+      <property name="format" value="\.clearValue\("/>
+      <property name="ignoreComments" value="true"/>
+      <property name="message" value="Please use withProperties approach instead." />
```
Comment (smiklosovic) [inline checkstyle_test.xml]: @bbotella please change the message to 

`Please use WithProperties in try-with-resources instead. See CASSANDRA-18453.`

File: test/distributed/org/apache/cassandra/distributed/test/MessageForwardingTest.java [hunk 2]
```diff
@@ -122,6 +122,7 @@ else if (traceEntry.activity.contains("Enqueuing forwarded write to "))
         finally
         {
             if (originalTraceTimeout == null)
+                // checkstyle: suppress below 'clearValueSystemPropertyUsage'
```
Comment (Mmuzaf) [inline test/distributed/org/apache/cassandra/distributed/test/MessageForwardingTest.java]: I think this case can also be handled by `WithProperties`, right?

File: test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java [hunk 3]
```diff
@@ -473,6 +473,7 @@ public static void withProperty(CassandraRelevantProperties prop, boolean value,
         withProperty(prop, Boolean.toString(value), r);
     }
 
+    // checkstyle: suppress below 'clearValueSystemPropertyUsage'
```
Comment (Mmuzaf) [inline test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java]: The `checkstyle: suppress below ` uses `offCommentFormat` here [1], which means that all the checks below this comment are suppressed. This makes the checks wildly inaccurate, as a developer can still use `clearValue` and get no errors on the build. 

Should we use `checkstyle: suppress nearby` for all such places?

[1] https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilter.html

File: test/unit/org/apache/cassandra/config/DatabaseDescriptorTest.java [hunk 4]
```diff
@@ -259,6 +259,7 @@ public void testInvalidPartitionPropertyOverride() throws Exception
         }
         finally
         {
+            // checkstyle: suppress below 'clearValueSystemPropertyUsage'
```
Comment (Mmuzaf) [inline test/unit/org/apache/cassandra/config/DatabaseDescriptorTest.java]: I think we can handle it instead, right?

File: test/unit/org/apache/cassandra/net/ConnectionTest.java [hunk 5]
```diff
@@ -651,6 +651,7 @@ public void testPendingOutboundConnectionUpdatesMessageVersionOnReconnectAttempt
         {
             MessagingService.instance().versions.set(FBUtilities.getBroadcastAddressAndPort(),
                                                      current_version);
+            // checkstyle: suppress below 'clearValueSystemPropertyUsage'
```
Comment (Mmuzaf) [inline test/unit/org/apache/cassandra/net/ConnectionTest.java]: The same as above, it looks like we can handle it :-)

File: test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java (hunk 3 above)
Comment (smiklosovic) [reply test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java]: sure, good idea.

File: test/unit/org/apache/cassandra/db/commitlog/CommitLogTest.java [hunk 6]
```diff
@@ -894,12 +891,11 @@ public void handleMutation(Mutation m, int size, int entryLocation, CommitLogDes
         }
     }
 
-    private void assertReplay(int expectedReplayedMutations, Runnable systemPropertySetter) throws Throwable
+    // private void assertReplay(int expectedReplayedMutations, Runnable systemPropertySetter) throws Throwable
```
Comment (Mmuzaf) [inline test/unit/org/apache/cassandra/db/commitlog/CommitLogTest.java]: I think we can remove this comment :-)

**Comment by Mmuzaf** [comment]
I'm +1 for these changes, leaving a small comment. I think it is common practice to fix something right before committing, so LGMT. @bbotella thank you very much for picking this up!

**Review by Mmuzaf — APPROVED** [review]

File: test/unit/org/apache/cassandra/db/commitlog/CommitLogTest.java (hunk 6 above)
Comment (bbotella) [reply test/unit/org/apache/cassandra/db/commitlog/CommitLogTest.java]: LOL yeah...!

File: test/distributed/org/apache/cassandra/distributed/test/FailingResponseDoesNotLogTest.java [hunk 7]
```diff
@@ -70,13 +71,13 @@ public static void beforeClassTopLevel() // need to make sure not to conflict wi
     @Test
     public void dispatcherErrorDoesNotLock() throws IOException
     {
-        CUSTOM_QUERY_HANDLER_CLASS.setString(AlwaysRejectErrorQueryHandler.class.getName());
-        try (Cluster cluster = Cluster.build(1)
-                                      .withConfig(c -> c.with(Feature.NATIVE_PROTOCOL, Feature.GOSSIP)
-                                                        .set("client_error_reporting_exclusions", ImmutableMap.of("subnets", Collections.singletonList("127.0.0.1")))
-                                      )
-                                      .start())
+        try (WithProperties properties = new WithProperties().set(CUSTOM_QUERY_HANDLER_CLASS, AlwaysRejectErrorQueryHandler.class.getName()))
```
Comment (jacek-lewandowski) [inline test/distributed/org/apache/cassandra/distributed/test/FailingResponseDoesNotLogTest.java]: Should cluster creation be included in try-with-resources?

File: test/distributed/org/apache/cassandra/distributed/test/NativeMixedVersionTest.java [hunk 8]
```diff
@@ -41,15 +42,15 @@ public void v4ConnectionCleansUpThreadLocalState() throws IOException
...
-        try (Cluster cluster = Cluster.build(1)
-                                      .withConfig(c ->
-                                                  c.with(Feature.values())
-                                                   .set("read_thresholds_enabled", true)
-                                                   .set("local_read_size_warn_threshold", "1KiB")
-                                      )
-                                      .start())
+        try (WithProperties properties = new WithProperties().set(IO_NETTY_EVENTLOOP_THREADS, 1))
         {
+            Cluster cluster = Cluster.build(1)
```
Comment (jacek-lewandowski) [inline test/distributed/org/apache/cassandra/distributed/test/NativeMixedVersionTest.java]: again.. leaves cluster unclosed

File: test/distributed/org/apache/cassandra/distributed/test/PaxosRepair2Test.java [hunk 9]
```diff
@@ -333,14 +335,12 @@ private static void assertLowBoundPurged(Cluster cluster)
...
-        AUTO_REPAIR_FREQUENCY_SECONDS.setInt(1);
-        DISABLE_PAXOS_AUTO_REPAIRS.setBoolean(true);
-        try (Cluster cluster = init(Cluster.create(3, cfg -> cfg
-                                                             .set("paxos_variant", "v2")
-                                                             .set("paxos_repair_enabled", true)
-                                                             .set("truncate_request_timeout_in_ms", 1000L)))
-        )
+        try (WithProperties properties = new WithProperties().set(AUTO_REPAIR_FREQUENCY_SECONDS, 1).set(DISABLE_PAXOS_AUTO_REPAIRS, true))
         {
+            Cluster cluster = init(Cluster.create(3, cfg -> cfg
```
Comment (jacek-lewandowski) [inline test/distributed/org/apache/cassandra/distributed/test/PaxosRepair2Test.java]: same

**Review by jacek-lewandowski — CHANGES_REQUESTED** [review]
Great to see that PR

I have two questions:
- is there any particular reason why the `Cluster` object was moved out from try-with-resources and left unclosed in several places?
- in the tests where clearing is suppressed - why just don't create WithProperties in Before/BeforeClass and close it in After/AfterClass correspondingly?

File: test/distributed/org/apache/cassandra/distributed/test/MessageForwardingTest.java [hunk 10]
```diff
@@ -122,7 +122,7 @@ else if (traceEntry.activity.contains("Enqueuing forwarded write to "))
         finally
         {
             if (originalTraceTimeout == null)
-                WAIT_FOR_TRACING_EVENTS_TIMEOUT_SECS.clearValue();
+                WAIT_FOR_TRACING_EVENTS_TIMEOUT_SECS.clearValue(); // checkstyle: suppress nearby 'clearValueSystemPropertyUsage'
```
Comment (jacek-lewandowski) [inline test/distributed/org/apache/cassandra/distributed/test/MessageForwardingTest.java]: why not use `WithProperties` in this class as well?

File: test/distributed/org/apache/cassandra/distributed/test/FailingResponseDoesNotLogTest.java [hunk 11]
```diff
@@ -70,13 +71,16 @@ public static void beforeClassTopLevel() // need to make sure not to conflict wi
...
     public void dispatcherErrorDoesNotLock() throws IOException
     {
-        CUSTOM_QUERY_HANDLER_CLASS.setString(AlwaysRejectErrorQueryHandler.class.getName());
-        try (Cluster cluster = Cluster.build(1)
-                                      .withConfig(c -> c.with(Feature.NATIVE_PROTOCOL, Feature.GOSSIP)
-                                                        .set("client_error_reporting_exclusions", ImmutableMap.of("subnets", Collections.singletonList("127.0.0.1")))
-                                      )
-                                      .start())
+        try (
+            WithProperties properties = new WithProperties().set(CUSTOM_QUERY_HANDLER_CLASS, AlwaysRejectErrorQueryHandler.class.getName());
```
Comment (jacek-lewandowski) [inline test/distributed/org/apache/cassandra/distributed/test/FailingResponseDoesNotLogTest.java]: nit: formatting  - should it be in the same line as try?

File: test/distributed/org/apache/cassandra/distributed/test/BootstrapBinaryDisabledTest.java [hunk 12]
```diff
@@ -62,7 +62,7 @@ public static void beforeClass() throws Throwable
     public static void afterClass()
     {
         if (originalResetBootstrapProgress == null)
-            RESET_BOOTSTRAP_PROGRESS.clearValue();
+            RESET_BOOTSTRAP_PROGRESS.clearValue(); // checkstyle: suppress nearby 'clearValueSystemPropertyUsage'
```
Comment (jacek-lewandowski) [inline test/distributed/org/apache/cassandra/distributed/test/BootstrapBinaryDisabledTest.java]: use `WithProperties` ?

File: test/distributed/org/apache/cassandra/distributed/test/ring/BootstrapTest.java [hunk 13]
```diff
@@ -84,7 +84,7 @@ public void afterTest()
     {
         CassandraRelevantProperties.MIGRATION_DELAY.setLong(savedMigrationDelay);
         if (originalResetBootstrapProgress == null)
-            RESET_BOOTSTRAP_PROGRESS.clearValue();
+            RESET_BOOTSTRAP_PROGRESS.clearValue(); // checkstyle: suppress nearby 'clearValueSystemPropertyUsage'
```
Comment (jacek-lewandowski) [inline test/distributed/org/apache/cassandra/distributed/test/ring/BootstrapTest.java]: Why not use `WithProperties`?

File: test/unit/org/apache/cassandra/config/CassandraRelevantPropertiesTest.java [hunk 14]
```diff
@@ -88,92 +79,65 @@ public void testBoolean()
     @Test
     public void testBoolean_null()
     {
-        try
+        try (WithProperties properties = new WithProperties())
         {
             TEST_CASSANDRA_RELEVANT_PROPERTIES.getBoolean();
```
Comment (jacek-lewandowski) [inline test/unit/org/apache/cassandra/config/CassandraRelevantPropertiesTest.java:84]: btw. there is missing assertion in this test case

File: test/unit/org/apache/cassandra/service/AbstractFilesystemOwnershipCheckTest.java [hunk 15]
```diff
@@ -220,7 +220,7 @@ public void checkEnabledButClusterPropertyIsEmpty()
     public void checkEnabledButClusterPropertyIsUnset()
     {
         Assume.assumeFalse(options.getConfig(check_filesystem_ownership).containsKey("ownership_token"));
-        CassandraRelevantProperties.FILE_SYSTEM_CHECK_OWNERSHIP_TOKEN.clearValue();
+        CassandraRelevantProperties.FILE_SYSTEM_CHECK_OWNERSHIP_TOKEN.clearValue(); // checkstyle: suppress nearby 'clearValueSystemPropertyUsage'
```
Comment (jacek-lewandowski) [inline test/unit/org/apache/cassandra/service/AbstractFilesystemOwnershipCheckTest.java:228]: WithProperties should be used in this class as well

**Review by jacek-lewandowski — CHANGES_REQUESTED** [review]

File: test/unit/org/apache/cassandra/config/CassandraRelevantPropertiesTest.java (hunk 14 above)
Comment (bbotella) [reply test/unit/org/apache/cassandra/config/CassandraRelevantPropertiesTest.java:84]: Will add it.

**Review by jacek-lewandowski — APPROVED** [review]
+1, Thanks @bbotella for considering my comments.

**Pull request closed without merging**